### ISSUE TITLE
Add search and search bar

### ISSUE THREAD

**Issue by vinsdragonis** [vinsdragonis/Project-Nexus#93]
With more posts created more frequently than before, it would be more convenient if we had a search bar through which we could search for specific posts based on their:
- Title
- Author
- Category

**Comment by Sam-Takahashi** [comment]
I'm willing to look into this. 
Do you mean to have a search bar for regular(typed) search queries, and another system for filtering by title, author, and category?

**Comment by vinsdragonis** [comment]
No, I meant that we can have a normal search bar that only produces relevant search results. Filtering posts can be done once the search results are displayed.

**Comment by vinsdragonis** [comment]
Let's ensure that some search results are displayed right below the search bar. And once the user clicks the search button, or hits the return key, we display all relevant results, keeping in mind that we should limit the number of posts displayed at a time with a `Load more` button to bring in more posts as and when required.

**Comment by Sam-Takahashi** [comment]
> Let's ensure that some search results are displayed right below the search bar. 

I'm unsure of exactly what you mean, can you please elaborate? 

Instead of a <code>Load more</code> button, can we paginate the results?

**Comment by vinsdragonis** [comment]
I mean, you can do that. I just suggested it since we already have such a feature implemented for the homepage. If pagination seems the right way to do it, you can use it.

**Comment by vinsdragonis** [comment]
This has been taken care of by @Sam-Takahashi in PR #110

**Comment by vikashjajra** [comment]
### The search bar code has created an app-breaking issue where it won't let the app initialize at all.
Here is how the live deployment at [https://projectnexus.netlify.app](https://projectnexus.netlify.app) looks
![Screenshot from 2022-08-29 10-39-35](https://user-images.githubusercontent.com/96776232/187127710-f1c185b7-b0aa-4076-9b34-4b9a3ca88320.png)

Here's what the console says
![Screenshot from 2022-08-29 10-40-22](https://user-images.githubusercontent.com/96776232/187127785-89f56ce1-842b-46d9-b99f-db2820e02814.png)
 
Here console shows an error in App.js:43 which looks like this
![Screenshot from 2022-08-29 10-40-55](https://user-images.githubusercontent.com/96776232/187127848-a658d74c-817b-4100-9768-9d178a2a9812.png)

I think the search needs a bit more refinement.

**Comment by vinsdragonis** [comment]
Would you like to take it up? @vikashjajra

**Comment by vikashjajra** [comment]
Sure... I have made a change that fixes the error for the moment in #113 , but the search works only on post titles now. I will try to fix it so the search can be performed on the author and the post content as well.

**Comment by Sam-Takahashi** [comment]
Hi, creator of the search bar here. 
It appears `item[key]` is undefined. Have there been any changes to the API/endpoint?
Let me know if the problem persists @vikashjajra and I'll look into it.

**Comment by vikashjajra** [comment]
> Hi, creator of the search bar here. It appears `item[key]` is undefined. Have there been any changes to the API/endpoint? Let me know if the problem persists @vikashjajra and I'll look into it.

There seem to be no changes in the APIs.

**Comment by vikashjajra** [comment]
> Hi, creator of the search bar here. It appears `item[key]` is undefined. Have there been any changes to the API/endpoint? Let me know if the problem persists @vikashjajra and I'll look into it.

I've been fiddling around with the code and found that when we remove the username from the keys array in App.js, it works fine. but adding it back recreates the problem.

**Comment by Sam-Takahashi** [comment]
> I've been fiddling around with the code and found that when we remove the username from the keys array in App.js, it works fine. but adding it back recreates the problem.

If all other keys in the array(other than `username`) work without issue, then there's a high possibility that the API schema has been changed(eg. `username` may have been removed or renamed).

**Comment by vinsdragonis** [comment]
@Sam-Takahashi you made a modification to the schema back during the profile update, right? There might a possibility that the issue came from there. 
Could you have a look at it?

**Comment by Sam-Takahashi** [comment]
The schema definitely has been changed.
<img width="1420" alt="image" src="https://user-images.githubusercontent.com/63477791/187201091-e1d4b6b8-ca68-416c-828e-9f185eed14c4.png">

**Comment by Sam-Takahashi** [comment]
We(Lydia and I) modified the `models/ProfilePage.js`, but we have not modified the `User` schema, which is causing this error.
<img width="837" alt="image" src="https://user-images.githubusercontent.com/63477791/187201500-249725d5-7a85-471a-b772-d909fc865474.png">

**Comment by vikashjajra** [comment]
A workaround will be needed for the username, in that case, because having the user's id would make more sense when making relationships. Also, I've been working on the upvotes system which now relies on the same relationships. The new json object for the post will be something like this:

    post:[
      post_data,
      user:[
        user_data
      ],
      votes:[
        an array of votes on the post
      ]
    ]

**Comment by Sam-Takahashi** [comment]
Before working on a workaround, shouldn't we investigate how the schema was changed? All the dummy data appears to have been wiped out as well, what was the reason for this?

**Comment by vinsdragonis** [comment]
@Sam-Takahashi could take a look at the changes you made from the previous update?
Far as I remember, that was the last time server files were changed. Maybe we missed out on something there.

**Comment by Sam-Takahashi** [comment]
> I've been working on the upvotes system which now relies on the same relationships.

I went over the changes I made, and there shouldn't be an issue if the Posts schema contains the `username` value.

I hate to point the finger, but @vikashjajra mentioned he's working on the upvote system. The current schema no longer contains a `username` and instead, it now contains a `votes` array.  I may be wrong in assuming this but perhaps @vikashjajra modified it for the upvote system?
Forgive me if I'm wrong.

**Comment by Sam-Takahashi** [comment]
https://github.com/vinsdragonis/Project-Nexus/issues/93#issuecomment-1230223201

**Comment by vinsdragonis** [comment]
I guess we need to be more careful making changes to the database hosted from now.
Better make a clone of the DB and locally test it. Regardless of where this issue started, we can't afford to test anything on live data ever again.

**Comment by vinsdragonis** [comment]
On changing the mongo URL on the server to a localhost version, it should automatically create a new local version of the DB after starting the server. Once tested, please share a screenshot of the working app to ensure everything works fine locally. That way, we can confirm that it can work the same way when hosted.

**Comment by vikashjajra** [comment]
> > I've been working on the upvotes system which now relies on the same relationships.
> 
> I went over the changes I made, and there shouldn't be an issue if the Posts schema contains the `username` value.
> 
> I hate to point the finger, but @vikashjajra mentioned he's working on the upvote system. The current schema no longer contains a `username` and instead, it now contains a `votes` array.  I may be wrong in assuming this but perhaps @vikashjajra modified it for the upvote system? https://github.com/vinsdragonis/Project-Nexus/issues/93#issuecomment-1230223201
> Forgive me if I'm wrong.
> 


Thanks for pointing this. you're absolutely right my friend. I didn't realise I was working on the hosted mongoURL instead of the local one.It is my fault.

**Comment by vinsdragonis** [comment]
@vikashjajra that is fine. Fortunately, the data was just to fill up space though it took some effort to put up. Please be careful about it next time, we don't want that to happen in a production environment.

**Comment by vikashjajra** [comment]
> @vikashjajra that is fine. Fortunately, the data was just to fill up space though it took some effort to put up. Please be careful about it next time, we don't want that to happen in a production environment.

I will be more cautious.
